### DISCUSSION TITLE
constraints: mass not found in expression

### DISCUSSION THREAD

**Discussion by dmrowan** [phoebe-project/phoebe2#618] [Questions]
Hello, 

I'm having some trouble using constraints and making sense of when I can flip back and forth between different sets of free parameters. 

Here is a simple of example that is similar to what I'm trying to do:

> b = phoebe.default_binary()
> b.flip_constraint('mass@primary', solve_for='sma@binary')
> b.set_value('mass@primary', 2)
> b.flip_constraint('sma@binary', solve_for='mass@primary')
> b.set_value('sma@binary', b['sma@binary'].value+1)
> b.flip_constraint('mass@primary', solve_for='sma@binary')
> b.set_value('mass@primary', 3)

In practice, I'm running optimization/MCMC between these steps. Here I am setting the values of mass and the sma to emulate 'kicking' the parameters and prevent falling into local minima between optimization steps or incorporating prior information from the literature (on asini, for example). 

Unfortunately, in this example above, I get the following error on line 6:
> KeyError: 'mass@primary@component was not found in expression'

My expectation was that I should be able to flip between sma and primary mass like this as many times as I like. 

I tried printing some additional information to see how line 2 and 6 differ. Before both of these lines I printed:
> print([var.get_parameter().twig for var in b['mass@primary@constraint']._vars])

Before the first call to b.flip_constraint('mass@primary', solve_for='sma@binary'), the output is

> ['sma@binary@orbit@component', 'period@binary@orbit@component', 'q@binary@orbit@component', 'mass@primary@star@component']

Before the second call to b.flip_constraint('mass@primary', solve_for='sma@binary'), the output is

> ['period@binary@orbit@component', 'q@binary@orbit@component', 'sma@binary@orbit@component']

Why does 'mass@primary@star@component' get removed during the flipping of constraints? Do you have any suggestions for what I should be doing differently so that I can run the original code snippet above?

Thanks!
Dom

**Comment by kecnry** [comment]
Hi Dom! - this looks like you're running into issue #342.  That issue has existed for a while (3 years 😬 ) but no one really ran into it in practice until now, so I didn't prioritize tracking down the cause.  I'll bump it up on the priority list, but I can't promise fixing it terribly soon.  You might be able to get around this by setting up all your parameters and datasets and then duplicating the bundle (with `b.copy()`) before flipping the constraints.  If you think this would work for what you need for now, but would need to adopt a solution from one bundle into the other copy, let me know and I can try to put together and example for how to make that work.

-Kyle

**Reply by dmrowan** [reply]
Hi Kyle, 

Thanks for the response! I should have searched through the issues tab -- sorry about that. I think copying the bundle, keeping the datasets and the current values of all parameters would work. Would this reset whatever is going wrong with the constrained parameters? Here is what I tried using b.copy():

> b = phoebe.default_binary()
> b.flip_constraint('mass@primary', solve_for='sma@binary')
> b.set_value('mass@primary', 2)
> b.flip_constraint('sma@binary', solve_for='mass@primary')
> b.set_value('sma@binary', b['sma@binary'].value+1)
> b2 = b.copy()
> b2.flip_constraint('mass@primary', solve_for='sma@binary')
> b2.set_value('mass@primary', 3)

I also tried copying the bundle before flipping for sma:
> b = phoebe.default_binary()
> b.flip_constraint('mass@primary', solve_for='sma@binary')
> b.set_value('mass@primary', 2)
> b2 = b.copy()
> b2.flip_constraint('sma@binary', solve_for='mass@primary')
> b2.set_value('sma@binary', b2['sma@binary'].value+1)
> b2.flip_constraint('mass@primary', solve_for='sma@binary')
> b2.set_value('mass@primary', 3)

Both cases return the same 'mass@primary@component was not found in expression' error as before. Is this how you were suggesting to use b.copy()? I also tried using _attach_params (with overwrite=True) to copy mass@primary@component and mass@primary@constraint. Here is what I did:

> b = phoebe.default_binary()
> b.flip_constraint('mass@primary', solve_for='sma@binary')
> b.set_value('mass@primary', 2)
> b.flip_constraint('sma@binary', solve_for='mass@primary')
> b.set_value('sma@binary', b['sma@binary'].value+1)
> 
> m1_component_copy = b['mass@primary@component'].copy()
> m1_component_copy.set_uniqueid(b['mass@primary@component'].uniqueid)
> m1_constraint_copy = b['mass@primary@constraint'].copy()
> m1_constraint_copy.set_uniqueid(b['mass@primary@constraint'].uniqueid)
> 
> b._attach_params([m1_component_copy], overwrite=True)
> b._attach_params([m1_constraint_copy], overwrite=True)
> 
> b.flip_constraint('mass@primary', solve_for='sma@binary')
> b.set_value('mass@primary', 3)

This seems to have worked... but I should emphasize I don't understand the 'uniqueid' system at all. My thought was that if the uniqueid for the primary mass is stored in any other constrained parameters, I need to make the new mass@primary@component and mass@primary@constraint have the same uniqueids as the original bundle.

Alternatively, I tried just updating the _vars attribute of the mass@primary@constraint so it doesn't lose mass@primary@component:

> b = phoebe.default_binary()
> original_vars = b['mass@primary@constraint']._vars
> b.flip_constraint('mass@primary', solve_for='sma@binary')
> b.set_value('mass@primary', 2)
> b.flip_constraint('sma@binary', solve_for='mass@primary')
> b.set_value('sma@binary', b['sma@binary'].value+1)
> b['mass@primary@constraint']._vars = original_vars
> b.flip_constraint('mass@primary', solve_for='sma@binary')
> b.set_value('mass@primary', 3)

This also lets me flip and set mass@primary again. Are there any reasons I should be careful adjusting the parameters this way? Using copy in some way is probably preferable, I just couldn't quite get it working. 

Cheers,
Dom

**Reply by kecnry** [reply]
Your hack at the end should work (I'm sorry that you had to dig that deep!).  I have an experimental fix for the underlying bug  if you want to try that out.  I'm hoping to get one other constraint-flipping bug fixed as well and then I'll release a bugfix for these asap.

Edit: fixed the other constraint bug and opened a PR for the bugfix: https://github.com/phoebe-project/phoebe2/pull/619
Edit 2: released as 2.4.4

Thanks for helping to track this down!

**Reply by dmrowan** [reply]
Great, I was able to flip the constraints as I wished and successfully run the emcee solver! Thanks again for your help!

Dom
